src/dpdkstat.c

```diff
@@ -0,0 +1,775 @@
+      if(g_configuration->eal_initialized) {
+        WARNING("dpdkstat-helper: no primary alive but EAL initialized:"
+              " quitting.\n");
+        dpdk_helper_exit(RESET);
+      }
+      g_configuration->helper_status = DPDK_HELPER_WAITING_ON_PRIMARY;
+      /* Back to start of while() - waiting for primary process */
+      continue;
+    }
+
+    if (!g_configuration->eal_initialized) {
+      /* Initialize EAL. */
+      int ret = dpdk_helper_init_eal();
+      if(ret != 0) {
+        WARNING("ERROR INITIALIZING EAL\n");
+        dpdk_helper_exit(RESET);
+      }
+    }
+
+    g_configuration->helper_status = DPDK_HELPER_ALIVE_SENDING_STATS;
+
+    uint8_t nb_ports = rte_eth_dev_count();
+    if (nb_ports == 0) {
+      DEBUG("dpdkstat-helper: No DPDK ports available. "
+              "Check bound devices to DPDK driver.\n");
+      return 0;
+    }
+
+    if (nb_ports > RTE_MAX_ETHPORTS)
+      nb_ports = RTE_MAX_ETHPORTS;
+
+    int len = 0, enabled_port_count = 0, num_xstats = 0;
+    for (uint8_t i = 0; i < nb_ports; i++) {
+      if (g_configuration->enabled_port_mask & (1 << i)) {
+        if(g_configuration->helper_action == DPDK_HELPER_ACTION_COUNT_STATS) {
+          len = rte_eth_xstats_get(i, NULL, 0);
+          if (len < 0) {
+            ERROR("dpdkstat-helper: Cannot get xstats count\n");
+            return -1;
+          }
+          num_xstats += len;
+          g_configuration->num_stats_in_port[enabled_port_count] = len;
+          enabled_port_count++;
+          continue;
+        } else {
+          len = g_configuration->num_stats_in_port[enabled_port_count];
+          g_configuration->port_read_time[enabled_port_count] = cdtime();
+          ret = rte_eth_xstats_get(i, &g_configuration->xstats + num_xstats,
+                                   g_configuration->num_stats_in_port[enabled_port_count]);
+          if (ret < 0 || ret != len) {
+            DEBUG("dpdkstat-helper: Error reading xstats on port %d len = %d\n",
+                  i, len);
+            return -1;
+          }
+          num_xstats += g_configuration->num_stats_in_port[enabled_port_count];
+          enabled_port_count++;
+        }
+      } /* if (enabled_port_mask) */
+    } /* for (nb_ports) */
+
+    if (g_configuration->helper_action == DPDK_HELPER_ACTION_COUNT_STATS) {
+      g_configuration->num_ports  = enabled_port_count;
+      g_configuration->num_xstats = num_xstats;
+      DEBUG("dpdkstat-helper ports: %d, num stats: %d\n",
+            g_configuration->num_ports, g_configuration->num_xstats);
+      /* Exit, allowing collectd to re-init SHM to the right size */
+      g_configuration->collectd_reinit_shm = REINIT_SHM;
+      dpdk_helper_exit(NO_RESET);
+    }
+    /* Now kick collectd send thread to send the stats */
+    int err = sem_post(&g_configuration->sema_stats_in_shm);
+    if (err)
+      ERROR("dpdkstat: error posting semaphore to helper %s\n", strerror(errno));
```

Comment:
Isn't this the helper code?

→ `sstrerror()`

Should this be a `WARNING()` or should the code not try to continue when the semaphore fails?